torch/csrc/autograd/engine.cpp

```diff
@@ -282,12 +286,10 @@ void Engine::thread_init(int device, const std::shared_ptr<ReadyQueue>& ready_qu
   init_local_ready_queue(ready_queue);
 
   std::shared_ptr<GraphTask> graph_task = nullptr;
-  thread_main(graph_task, /* reentrant_thread */ false);
-  // Notify about shutdown
-  {
-    std::unique_lock<std::mutex> lk(non_reentrant_device_thread_mutex_);
```

Comment:
Was it a bug that this was not properly protected by the if statement?

Reply:
Not really a bug since we mostly use PyThonEngine and it was taken care of there (since we always pass false here). I just felt it made sense to have this API do the right thing standalone too.